src/System.Management.Automation/engine/PSVersionInfo.cs

```diff
@@ -85,40 +85,60 @@ static PSVersionInfo()
             //
             // The strings can be one of the following format examples:
             //    when powershell is built from a commit:
-            //      productVersion = '6.0.0-beta.7 Commits: 29 SHA: 52c6b...' convert to GitCommitId = 'v6.0.0-beta.7-29-g52c6b...'
-            //                                                                           PSVersion   = '6.0.0-beta.7'
+            //      productVersion = '7.3.0-preview.7 Commits: 29 SHA: 52c6b...' convert to GitCommitId = 'v7.3.0-preview.7-29-g52c6b...'
+            //                                                                           PSVersion   = '7.3.0-preview.7'
             //    when powershell is built from a release tag:
-            //      productVersion = '6.0.0-beta.7 SHA: f1ec9...'             convert to GitCommitId = 'v6.0.0-beta.7'
-            //                                                                           PSVersion   = '6.0.0-beta.7'
+            //      productVersion = '7.3.0-preview.7 SHA: f1ec9...'             convert to GitCommitId = 'v7.3.0-preview.7'
+            //                                                                           PSVersion   = '7.3.0-preview.7'
             //    when powershell is built from a release tag for RTM:
-            //      productVersion = '6.0.0 SHA: f1ec9...'                    convert to GitCommitId = 'v6.0.0'
-            //                                                                           PSVersion   = '6.0.0'
-            string rawGitCommitId;
-            string mainVersion = ProductVersion.Substring(0, ProductVersion.IndexOf(' '));
+            //      productVersion = '7.3.0 SHA: f1ec9...'                    convert to GitCommitId = 'v7.3.0'
+            //                                                                           PSVersion   = '7.3.0'
+            int index = ProductVersion.IndexOf(' ');
+            ReadOnlySpan<char> mainVersion = ProductVersion.AsSpan(0, index);
 
-            if (ProductVersion.Contains(" Commits: "))
-            {
-                rawGitCommitId = ProductVersion.Replace(" Commits: ", "-").Replace(" SHA: ", "-g");
-            }
-            else
-            {
-                rawGitCommitId = mainVersion;
-            }
+            string rawGitCommitId = ProductVersion.AsSpan(index).StartsWith(" Commits: ")
+                ? "v" + ProductVersion.Replace(" Commits: ", "-").Replace(" SHA: ", "-g")
+                : string.Concat("v".AsSpan(), mainVersion);
 
-            s_psSemVersion = new SemanticVersion(mainVersion);
-            s_psVersion = (Version)s_psSemVersion;
+            var r = FastParsePSVersion(mainVersion);
```

Comment:
Would it make sense to turn this into a source generator since it'll be known at compile time?

Reply:
Ah I see the other PR

Reply:
Yeah, absolutely want to enable the source generator PR along with this one. I will do some measurement with changes from both PRs.

Reply:
I think it's doable. Will submit a separate PR for that.

Reply:
With a source generator, would we need to parse anything at all? Or could we generate a direct `new Version(` statement?

In other words, the source generator will also have access to the product version, so couldn't we generate something like this?

```csharp
// PSVersionInfo.generated.cs
public static partial class PSVersionInfo
{
    private static readonly Version s_psVersion = new(7, 3, 0);
}
```

Reply:
It is doing that now, the source generator PR has been updated:
https://github.com/PowerShell/PowerShell/pull/15603/files